### PR TITLE
Order by aggregation not present in selection

### PR DESCRIPTION
**Pull request by npawar**
Supporting queries where `order by` is on aggregation, but that aggregation is not defined in the selection:
```
select subject from transcript group by subject order by count(*), sum(score) desc
```

This is equivalent to 
```
select subject, count(*) from transcript group by subject order by count(*), sum(score) desc
```
but the return response should not contain count(*). This is a valid SQL query and we saw that we are only looking at expressions in the select clause and not in order by. This PR adds the functionality to check for order by expression as well to decide if its an aggregation query.

### CODE REVIEW

**Comment by fx19880617** [comment]
Can we add some query for integration tests?

File: pinot-common/src/main/java/org/apache/pinot/pql/parsers/PinotQuery2BrokerRequestConverter.java [hunk 1]
```diff
@@ -83,15 +83,30 @@ private void convertOrderBy(PinotQuery pinotQuery, BrokerRequest brokerRequest)
...
       sortSequenceList.add(selectionSort);
+
+      if (expression.getType().equals(ExpressionType.FUNCTION) && FunctionDefinitionRegistry
+          .isAggFunc(expression.getFunctionCall().getOperator())) {
+        AggregationInfo aggInfo = buildAggregationInfo(expression.getFunctionCall());
+        if (aggregationInfoList == null) {
+          aggregationInfoList = new ArrayList<>();
+          aggregationInfoList.add(aggInfo);
+          brokerRequest.setAggregationsInfo(aggregationInfoList);
+        } else if (!aggregationInfoList.contains(aggInfo)) {
```
Comment (kishoreg) [inline pinot-common/src/main/java/org/apache/pinot/pql/parsers/PinotQuery2BrokerRequestConverter.java:104]: Does this call work?

File: pinot-core/src/main/java/org/apache/pinot/core/query/request/context/utils/QueryContextUtils.java [hunk 2]
```diff
@@ -75,6 +78,14 @@ public static boolean isAggregationQuery(QueryContext query) {
         return true;
       }
     }
+    if(query.getOrderByExpressions() != null) {
```
Comment (kishoreg) [inline pinot-core/src/main/java/org/apache/pinot/core/query/request/context/utils/QueryContextUtils.java:81]: Since we are adding to aggregationInfo already, we should probably check for aggregation info

File: pinot-common/src/main/java/org/apache/pinot/pql/parsers/PinotQuery2BrokerRequestConverter.java (hunk 1 above)
Comment (mayankshriv) [reply pinot-common/src/main/java/org/apache/pinot/pql/parsers/PinotQuery2BrokerRequestConverter.java:104]: Isn't this really an aggregation group-by query (or at least needs to be executed like it)? If so, how does the aggregation actually work here? Will `QueryContextUtils.isAggregationQuery`  return true in this case?

**Comment by npawar** [comment]
> Can we add some query for integration tests?

Added some tests in CalciteSQlCompilertest. Our integration tests are still calling pql endpoint, so these cannot be added there

File: pinot-common/src/main/java/org/apache/pinot/pql/parsers/PinotQuery2BrokerRequestConverter.java (hunk 1 above)
Comment (npawar) [reply pinot-common/src/main/java/org/apache/pinot/pql/parsers/PinotQuery2BrokerRequestConverter.java:104]: @kishoreg  yes it works.
@mayankshriv I've modified `QueryContextUtils.isAggregationQuery` also in same PR, and so it does return true. Yes it gets treated as a aggregation-group by query, and finally only the columns from the selection are returned

File: pinot-core/src/main/java/org/apache/pinot/core/query/request/context/utils/QueryContextUtils.java (hunk 2 above)
Comment (npawar) [reply pinot-core/src/main/java/org/apache/pinot/core/query/request/context/utils/QueryContextUtils.java:81]: there is no concept of aggregations in the QueryContext

Comment (kishoreg) [reply pinot-core/src/main/java/org/apache/pinot/core/query/request/context/utils/QueryContextUtils.java:81]: My bad, I forgot we moved to querycontext already.

**Comment by siddharthteotia** [comment]
@npawar , Can we please add end-to-end query execution tests? I don't mean integration tests but tests in one of the files that extend BaseSingleValueQueriesTest and BaseMultiValueQueriesTest. All e2e query exec tests that test out the entire path from broker are there.

**Comment by npawar** [comment]
> @npawar , Can we please add end-to-end query execution tests? I don't mean integration tests but tests in one of the files that extend BaseSingleValueQueriesTest and BaseMultiValueQueriesTest. All e2e query exec tests that test out the entire path from broker are there.

Added @siddharthteotia

**Comment by Jackie-Jiang** [comment]
Are we planning to support this for SQL only? In that case, we should rewrite the query (add all the order-by expressions to the selection list) for PinotQuery instead of in the converter and only have the rewrite in BrokerRequest. This will cause inconsistency between PinotQuery and BrokerRequest, which will make BrokerRequest deprecation harder.

File: pinot-core/src/test/java/org/apache/pinot/queries/InterSegmentOrderBySingleValueQueriesTest.java [hunk 3]
```diff
@@ -165,7 +162,7 @@ public void testQueryOptions() {
         new Object[]{query, results, numDocsScanned, numEntriesScannedInFilter, numEntriesScannedPostFilter, numTotalDocs, dataSchema});
 
     // order by one of the group by columns DESC
-    query = "SELECT sum(column1) FROM testTable GROUP BY column11 ORDER BY column11 DESC";
+    query = "SELECT column11, sum(column1) FROM testTable GROUP BY column11 ORDER BY column11 DESC";
```
Comment (siddharthteotia) [inline pinot-core/src/test/java/org/apache/pinot/queries/InterSegmentOrderBySingleValueQueriesTest.java:165]: Why is this change needed?

Comment (npawar) [reply pinot-core/src/test/java/org/apache/pinot/queries/InterSegmentOrderBySingleValueQueriesTest.java:165]: this was using the pql endpoint with options because we were not ready for the sql parser. i changed it to use SQL endpoint. In sql endpoint if column is expecetd in results, then it needs to be in the selection

File: pinot-core/src/test/java/org/apache/pinot/queries/InterSegmentOrderBySingleValueQueriesTest.java [hunk 4]
```diff
@@ -186,7 +183,7 @@ public void testQueryOptions() {
         new Object[]{query, results, numDocsScanned, numEntriesScannedInFilter, numEntriesScannedPostFilter, numTotalDocs, dataSchema});
 
     // group by 2 dimensions, order by both, tie breaker
-    query = "SELECT SUM(column1) FROM testTable GROUP BY column11, column12 ORDER BY column11, column12";
+    query = "SELECT column11, column12, SUM(column1) FROM testTable GROUP BY column11, column12 ORDER BY column11, column12";
```
Comment (siddharthteotia) [inline pinot-core/src/test/java/org/apache/pinot/queries/InterSegmentOrderBySingleValueQueriesTest.java:186]: I am not sure why this change is needed.

File: pinot-common/src/test/java/org/apache/pinot/sql/parsers/CalciteSqlCompilerTest.java [hunk 5]
```diff
@@ -750,6 +753,136 @@ public void testSelectionTransformFunction() {
...
+    aggregationInfos = brokerRequest.getAggregationsInfo();
+    Assert.assertEquals(aggregationInfos.size(), 1);
+    Assert.assertEquals(aggregationInfos.get(0).getAggregationType(), AggregationFunctionType.COUNT.toString());
+    Assert.assertEquals(aggregationInfos.get(0).getExpressions().get(0), "*");
+    orderBy = brokerRequest.getOrderBy();
+    Assert.assertEquals(orderBy.size(), 1);
+    Assert.assertEquals(orderBy.get(0).getColumn(), "count(*)");
+
+    // group by not provided in selection
+    sql = "SELECT COUNT(*) FROM myTable GROUP BY foo ORDER BY foo";
```
Comment (siddharthteotia) [inline pinot-common/src/test/java/org/apache/pinot/sql/parsers/CalciteSqlCompilerTest.java:804]: I thought at some point we had made a decision that SQL GROUP BY would require the user to explicitly specify the GROUP BY columns in the select list and we were ok with users changing their queries to add group by columns to the select list. I know PQL doesn't have that restriction. Are we relaxing this restriction on SQL side as well?

File: pinot-core/src/test/java/org/apache/pinot/queries/InterSegmentOrderByMultiValueQueriesTest.java [hunk 6]
```diff
@@ -38,10 +38,7 @@
   @Test(dataProvider = "orderByDataProvider")
   public void testGroupByOrderByMVSQLResults(String query, List<Object[]> expectedResults,
       long expectedNumEntriesScannedPostFilter, DataSchema expectedDataSchema) {
-    Map<String, String> queryOptions = new HashMap<>(2);
```
Comment (siddharthteotia) [inline pinot-core/src/test/java/org/apache/pinot/queries/InterSegmentOrderByMultiValueQueriesTest.java:41]: I would recommend to not change existing unit tests to accommodate new tests. It will reduce the coverage like in this case we were testing out PQL execution path and we changed to test SQL. 

We should simply add a new unit test.

Comment (npawar) [reply pinot-core/src/test/java/org/apache/pinot/queries/InterSegmentOrderByMultiValueQueriesTest.java:41]: all other tests use either full pql or full sql. this middle ground was added by me when we were not ready.i dont see why we should keep this half approach, now that we have the new endpoint ready

File: pinot-common/src/test/java/org/apache/pinot/sql/parsers/CalciteSqlCompilerTest.java (hunk 5 above)
Comment (npawar) [reply pinot-common/src/test/java/org/apache/pinot/sql/parsers/CalciteSqlCompilerTest.java:804]: we didn't relax any restriction. the columns in the result will still be only what is specified in the select

File: pinot-core/src/test/java/org/apache/pinot/queries/InterSegmentOrderBySingleValueQueriesTest.java (hunk 3 above)
Comment (siddharthteotia) [reply pinot-core/src/test/java/org/apache/pinot/queries/InterSegmentOrderBySingleValueQueriesTest.java:165]: We should not change existing unit tests. I suggest to append to the unit tests without changing the existing queries in existing tests otherwise we will end up reducing the coverage.

Comment (siddharthteotia) [reply pinot-core/src/test/java/org/apache/pinot/queries/InterSegmentOrderBySingleValueQueriesTest.java:165]: We should simply add two unit tests -- one for PQL path and one for SQL path.

Comment (npawar) [reply pinot-core/src/test/java/org/apache/pinot/queries/InterSegmentOrderBySingleValueQueriesTest.java:165]: there is a test for pql path in the same file. this one should always have been sql path test

**Comment by npawar** [comment]
closed this PR since @Jackie-Jiang fixed this in https://github.com/apache/incubator-pinot/pull/5637 and the tests have been moved to that PR
